overlord/servicestate/quota_control.go

```diff
@@ -70,51 +70,49 @@ func ensureSnapServicesForGroup(st *state.State, grp *quota.Group, allGrps map[s
 	return wrappers.EnsureSnapServices(snapSvcMap, ensureOpts, nil, progress.Null)
 }
 
-func validateSnapForAddingToGroup(st *state.State, name string, allGrps map[string]*quota.Group) error {
-	// validate that the snap exists
-	_, err := snapstate.CurrentInfo(st, name)
-	if err != nil {
-		return err
-	}
+func validateSnapForAddingToGroup(st *state.State, snaps []string, group string, allGrps map[string]*quota.Group) error {
+	for _, name := range snaps {
+		// validate that the snap exists
+		_, err := snapstate.CurrentInfo(st, name)
+		if err != nil {
+			return fmt.Errorf("cannot use snap %q in group %q: %v", name, group, err)
+		}
 
-	// check that the snap is not already in a group
-	for _, grp := range allGrps {
-		if strutil.ListContains(grp.Snaps, name) {
-			return fmt.Errorf("snap already in quota group %q", grp.Name)
+		// check that the snap is not already in a group
+		for _, grp := range allGrps {
+			if strutil.ListContains(grp.Snaps, name) {
+				return fmt.Errorf("cannot add snap %q to group %q: snap already in quota group %q", name, group, grp.Name)
+			}
 		}
 	}
+
 	return nil
 }
 
 // CreateQuota attempts to create the specified quota group with the specified
 // snaps in it.
 // TODO: should this use something like QuotaGroupUpdate with fewer fields?
 func CreateQuota(st *state.State, name string, parentName string, snaps []string, memoryLimit quantity.Size) error {
-	st.Lock()
```

Comment:
locking goes to the caller then?

Reply:
Yeah, this is for consistency with similar methods in other packages.